### PR TITLE
Remove compiler passes and use default-index-method in tags for ContentTypeResolver and DataProviderResolver

### LINKED ISSUE #18: Remove compiler passes and use index_by in tags for ContentTypeResolver and DataProviderResolver
Currently tagged services `sulu_headless.content_resolver` and `sulu_headless.data_provider_resolver` for ContentTypeResolver and DataProviderResolver are 'named' by calling `getContentType` or `getDataProvider` on it and then injected in the service that needs them named.

See: https://github.com/sulu/SuluHeadlessBundle/blob/master/DependencyInjection/ContentTypeResolverCompilerPass.php#L34 and https://github.com/sulu/SuluHeadlessBundle/blob/master/DependencyInjection/DataProviderResolverCompilerPass.php#L34

This can be replaced by utilizing indexed and tagged service collections in symfony, see https://symfony.com/blog/new-in-symfony-4-3-indexed-and-tagged-service-collections

The implementation in the tagged services can be the same if we define default_index_method correctly.

I'm willing to make a PR for this the coming week(s).

### PR DESCRIPTION
**Pull request by reyostallenberg**
fixes #18

### CODE REVIEW

File: Content/ContentResolver.php [hunk 1]
```diff
@@ -23,12 +24,9 @@ class ContentResolver implements ContentResolverInterface
      */
     private $resolvers;
 
-    /**
-     * @param ContentTypeResolverInterface[] $resolvers
-     */
-    public function __construct(array $resolvers)
+    public function __construct(RewindableGenerator $resolvers)
```
Comment (nnatter) [inline Content/ContentResolver.php]: do you need to use `RewindableGenerator` here? how did you find the type? 🙂 

if not, would prefer to use `iterable` as described in the docs. that should make testing easier too: https://symfony.com/doc/4.4/service_container/tags.html#reference-tagged-services

File: Tests/Unit/Content/ContentResolverTest.php [hunk 2]
```diff
@@ -39,9 +40,12 @@ protected function setUp(): void
         $this->mediaSelectionResolver = $this->prophesize(ContentTypeResolverInterface::class);
 
         $this->contentResolver = new ContentResolver(
-            [
-                'media_selection' => $this->mediaSelectionResolver->reveal(),
-            ]
+            new RewindableGenerator(
```
Comment (nnatter) [inline Tests/Unit/Content/ContentResolverTest.php]: you probably do not need to change this when using `iterable` as type of the argument above

File: Tests/Unit/Content/ContentTypeResolver/SmartContentResolverTest.php [hunk 3]
```diff
@@ -79,7 +80,12 @@ protected function setUp(): void
         $this->targetGroupStore = $this->prophesize(TargetGroupStoreInterface::class);
 
         $this->smartContentResolver = new SmartContentResolver(
-            ['media' => $this->mediaProviderResolver->reveal()],
+            new RewindableGenerator(
```
Comment (nnatter) [inline Tests/Unit/Content/ContentTypeResolver/SmartContentResolverTest.php]: you probably do not need to change this when using `iterable` as type of the argument above

**Comment by reyostallenberg** [comment]
It has to be a `Traversable` for the `iterator_to_array` function, or all the calls on the iterables need to be changed, or you'll get `Cannot use object of type Symfony\\Component\\DependencyInjection\\Argument\\RewindableGenerator as array`

File: Content/ContentResolver.php (hunk 1 above)
Comment (reyostallenberg) [reply Content/ContentResolver.php]: It is not really possible

File: Tests/Unit/Content/ContentResolverTest.php (hunk 2 above)
Comment (reyostallenberg) [reply Tests/Unit/Content/ContentResolverTest.php]: It is not really possible without

File: Tests/Unit/Content/ContentTypeResolver/SmartContentResolverTest.php (hunk 3 above)
Comment (reyostallenberg) [reply Tests/Unit/Content/ContentTypeResolver/SmartContentResolverTest.php]: It is not really possible without

File: Content/ContentResolver.php [hunk 4]
```diff
@@ -23,12 +24,9 @@ class ContentResolver implements ContentResolverInterface
...
     {
-        $this->resolvers = $resolvers;
+        $this->resolvers = iterator_to_array($resolvers);
```
Comment (wachterjohannes) [inline Content/ContentResolver.php]: is this really needed? i would love to see that we does not need this - else we wont have to resolve the services if not needed

**Comment by alexander-schranz** [comment]
@reyostallenberg would typehint then the constructor with `Traversable` in the tests its ok for me to use `RewindableGenerator` what do you think @nnatter

File: Content/ContentResolver.php (hunk 4 above)
Comment (alexander-schranz) [reply Content/ContentResolver.php]: @wachterjohannes the services will not resolved aslong as we call not a function on them. So iterator_to_array will not have a effect and is at current state the only way to convert the generator to keys and is the [documented](https://symfony.com/blog/new-in-symfony-4-3-indexed-and-tagged-service-collections) way by symfony.

Comment (wachterjohannes) [reply Content/ContentResolver.php]: OK then everything is good

**Comment by nnatter** [comment]
@alexander-schranz @reyostallenberg Does this mean that code of the [announcement blog post](https://symfony.com/blog/new-in-symfony-4-3-indexed-and-tagged-service-collections) is not working? if yes, i would prefer `Traversable` because `RewindableGenerator` is [tagged as internal](https://github.com/symfony/dependency-injection/blob/master/Argument/RewindableGenerator.php#L15).

**Comment by alexander-schranz** [comment]
@nnatter the code works as symfony will give a `RewindableGenerator`  which is a `Traversable` and a `Traversable` is `iterable`. But you can not use the class and just give a iterable like an array to it in the tests you would can use instead a `ArrayIterator`.

So the tests I would change to `new ArrayIterator(...);` and the constructor should be typehinted with `Traversable`.

**Comment by nnatter** [comment]
@alexander-schranz Alright thanks for the explanation! `Traversable` as typehint for the service and `ArrayIterator` in the tests sounds good 👍

**Comment by reyostallenberg** [comment]
I think this is the most optimal solution

**Comment by nnatter** [comment]
Looks great - thanks a lot for your effort! 🙂

**Pull request merged**